File: c++ version/101-200/130-surrounded-regions.cpp
vector<int> directions{-1, 0, 1, 0, -1};
void solve(vector<vector<char>>& board) {
    int m = board.size(), n = board[0].size();
    vector<vector<bool>> canFlip(m, vector<bool>(n, true));
    
    for (int i = 0; i < m; ++i) {
        dfs(i, 0, board, canFlip);
        dfs(i, n-1, board, canFlip);
    }

    for (int j = 0; j < n; ++j) {
        dfs(0, j, board, canFlip);
        dfs(m-1, j, board, canFlip);
    }

    for (int i = 0; i < m; ++i) {
        for (int j = 0; j < n; ++j) {
            if (board[i][j] == 'O' && canFlip[i][j])
                board[i][j] = 'X';
        }
    }
}

void dfs(int x, int y, vector<vector<char>> &board, vector<vector<bool>> &canFlip) {
    if (board[x][y] == 'X' || !canFlip[x][y]) {
        return;
    }
    canFlip[x][y] = false;
    for (int k = 0; k < 4; ++k) {
        int row = x + directions[k], col = y + directions[k+1];
        if (row >= 0 && row < board.size() && col >= 0 && col < board[0].size()) {
            dfs(row, col, board, canFlip);
        }
    }
}

File: python version/130-surrounded-regions.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        m = len(board)
        n = len(board[0])
        directions = [-1, 0, 1, 0, -1]
        visited = [[False] * n for _ in range(m)]

        def dfs(i, j):
            if i<0 or i>=m or j<0 or j>=n: return
            if board[i][j] == 'X' or visited[i][j]: return
            visited[i][j] = True
            for k in range(4):
                dfs(i+directions[k], j+directions[k+1])
        
        for j in range(n):
            if board[0][j] == 'O' and visited[0][j] == False:
                dfs(0, j)
            if board[m-1][j] == 'O' and visited[m-1][j] == False:
                dfs(m-1, j)

        for i in range(m):
            if board[i][0] == 'O' and visited[i][0] == False:
                dfs(i, 0)
            if board[i][n-1] == 'O' and visited[i][n-1] == False:
                dfs(i, n-1)


        for i in range(m):
            for j in range(n):
                if board[i][j] == 'O' and visited[i][j] == False:
                    board[i][j] = 'X'